### PR TITLE
feat: Add initial objects for LLM events

### LINKED ISSUE #223: Create LLM event types
To facilitate tracking new events during AWS Bedrock instrumentation, we need to create the various LLM event types : LlmEmbedding, LlmChatCompletionSummary, LlmChatCompletionMessage, and LlmError.

The data types are defined in [Node.js AWS Bedrock CDD](https://newrelic.atlassian.net/wiki/spaces/INST/pages/3101950826/Node.js+AWS+Bedrock+CDD).  Please note we will need different parsers for handling the LlmChatCompletionSummary and LlmChatCompletionMessage for the various supported models.

**Note**: If models are unknown do not try to add events for unsupported models

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR is to resolve #223.

I have left out a lot of defensive checks because I'm not clear how the objects will be used at the moment. I've tried to implement the CDD as closely as possible in such a way as to provide a baseline that can be tweaked and completed as the actual instrumentation is written.

Since some of the class syntax is newer than the ECMAScript version configured for the linter on the main branch, this PR also updates all of the linting tooling.

### CODE REVIEW

File: lib/llm/event.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+   * stored in the current transaction.
+   *
+   * @param {object} agent The New Relic agent that provides access to the
+   * transaction.
+   *
+   * @returns {string}
+   */
+  conversationId(agent) {
+    const tx = agent.tracer.getTransaction()
+    const attrs = tx?.trace?.custom.get('TODO')
```
Comment (bizob2828) [inline lib/llm/event.js:162]: hmm so this is a bitwise operator in `lib/config/attributes-filter.js` within agent.

File: lib/llm/chat-completion-summary.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+   * @returns {Promise<LlmChatCompletionSummary>}
+   */
+  static async create(params = createDefaultParams) {
+    const { awsClient } = params
+    const event = new LlmChatCompletionSummary(params)
+    await event.addKeyDigits(awsClient)
+    return event
+  }
+
+  #setMaxTokens() {
```
Comment (bizob2828) [inline lib/llm/chat-completion-summary.js:98]: I think these key accessors need to be more defensive just in case

File: lib/llm/event.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    this.api_key_last_four_digits = null
+
+    this['response.model'] = this.invokeInput?.modelId
+    this['request.model'] = this.invokeInput?.modelId
+    this['request.max_tokens'] = null
+
+    this.duration = -1
+    this.content = null
+    this.role = null
+    this.sequence = -1
```
Comment (bizob2828) [inline lib/llm/event.js:81]: this is passed in as the index on creating of all the chat completion messages

**Review by bizob2828 — CHANGES_REQUESTED** [review]
this will prob take a few iterations.  Here's a [commit](https://github.com/bizob2828/node-newrelic-aws-sdk/commit/4ff81b8bdefe911eeaa528a46aff74ec653c62e6) that shows instrumentation in action and an attempt to create the relevant chat completion messages/summary.

this is also a [commit](https://github.com/bizob2828/node-newrelic-aws-sdk/commit/5dd2801dd51d0675b9c2177255c91ba4ade7178f) where i attempted to start fixing but got overwhelmed.  Overall it's close aside from the execution of creating chat completion messages.  they need to be done in a loop to pass in the index which i linked in the comments of the commits above.  we can sync up in morning as well

File: lib/llm/chat-completion-message.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+   * added. Specifically, it gets the credentials from the client and adds the
+   * last four digits of the access key to the message.
+   *
+   * @param {LlmChatCompletionCreateParams} params
+   *
+   * @returns {Promise<LlmChatCompletionMessage>}
+   */
+  static async create(params = createDefaultParams) {
+    const { awsClient } = params
+    const event = new LlmChatCompletionMessage(params)
```
Comment (bizob2828) [inline lib/llm/chat-completion-message.js:80]: this only exists on summary and embedding

File: lib/llm/event.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  /**
+   * Create a new LlmEvent with AWS specific information added. Specifically,
+   * it gets the credentials from the client and adds the last four digits of
+   * the access key to the event.
+   *
+   * @param {LlmEventCreateParams} params
+   *
+   * @returns {Promise<LlmEvent>}
+   */
+  static async create(params = createDefaultParams) {
```
Comment (bizob2828) [inline lib/llm/event.js:131]: this only exists on summary and embedding

File: lib/llm/chat-completion-message.js [hunk 6]
```diff
@@ -66,20 +57,9 @@ class LlmChatCompletionMessage extends LlmEvent {
...
-   * @returns {Promise<LlmChatCompletionMessage>}
-   */
-  static async create(params = createDefaultParams) {
-    const { awsClient } = params
-    const event = new LlmChatCompletionMessage(params)
-    await event.addKeyDigits(awsClient)
-    return event
+  #setIsResponse(message, response) {
+    // TODO: determine an algorithm for this
+    return Object.keys(response).length > 0
```
Comment (bizob2828) [inline lib/llm/chat-completion-message.js:62]: the algorithm in openai is the first response content matches the message content: https://github.com/newrelic/node-newrelic/blob/main/lib/llm-events/openai/chat-completion-message.js#L18

**Review by bizob2828 — CHANGES_REQUESTED** [review]
this is much better. i'm glad after back and forth we landed on some command and response parser separate from the llm messages, nice work. 

I have a patch of fixes/suggestions from using it
[changes.patch](https://github.com/newrelic/node-newrelic-aws-sdk/files/13835724/changes.patch)

Lastly, can I ask that you create an `lib/llm/index.js` and export all the classes as named exports so i can just destructure it all from `const { LlmChatCompletionMessage,... } = require('../llm/')`

File: lib/llm/chat-completion-summary.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    const upt = 'response.usage.prompt_tokens'
+    const uct = 'response.usage.completion_tokens'
+    const cfr = 'response.choices.finish_reason'
+    const rt = 'request.temperature'
+
+    const cmd = this.bedrockCommand
+    this[uct] = parseInt(this.bedrockResponse.headers['x-amzn-bedrock-output-token-count'] || 0, 10)
+    this[upt] = parseInt(this.bedrockResponse.headers['x-amzn-bedrock-input-token-count'] || 0, 10)
+    this[utt] = this[upt] + this[uct]
+    this[cfr] = cmd.finishReason
```
Comment (bizob2828) [inline lib/llm/chat-completion-summary.js:42]:
```suggestion
    this[cfr] = this.bedrockResponse.finishReason
```

File: lib/llm/bedrock-response.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  get finishReason() {
+    const cmd = this.#command
+    let result
+    if (cmd.isTitan() === true) {
+      result = this.#parsedBody.results?.[0]?.completionReason
+    } else if (cmd.isClaude() === true) {
+      result = this.#parsedBody.stop_reason
+    } else if (cmd.isAi21() === true) {
+      result = this.#parsedBody.results?.[0]?.finishReason.reason
```
Comment (bizob2828) [inline lib/llm/bedrock-response.js:63]:
```suggestion
      result = this.#parsedBody.completions?.[0]?.finishReason.reason
```

**Comment by jsumners-nr** [comment]
For easier reference:

```diff
diff --git a/lib/llm/bedrock-command.js b/lib/llm/bedrock-command.js
index 8548581..bf397bf 100644
--- a/lib/llm/bedrock-command.js
+++ b/lib/llm/bedrock-command.js
@@ -14,13 +14,12 @@ class BedrockCommand {
   #modelId
 
   /**
-   * @param {object} params
-   * @param {object} params.invokeCommand The InvokeModelCommand or
+   * @param {object} input The InvokeModelCommand or
    * InvokeModelWithResponseStreamCommand instance that is used for the
    * conversation.
    */
-  constructor({ invokeCommand }) {
-    this.#input = invokeCommand.input
+  constructor(input) {
+    this.#input = input
     this.#body = JSON.parse(this.#input.body)
     this.#modelId = this.#input.modelId.toLowerCase()
   }
@@ -32,10 +31,10 @@ class BedrockCommand {
    *
    * @returns {boolean}
    */
-  static isEmbeddingModel(modelId) {
+  get modelType() {
     // At the moment, this is a simple check. If Amazon ever introduces a
     // complex identifier, we can implement a more complicated check.
-    return modelId.toLowerCase().includes('embed')
+    return this.#modelId.toLowerCase().includes('embed') ? 'embedding' : 'completion'
   }
 
   /**
diff --git a/lib/llm/bedrock-response.js b/lib/llm/bedrock-response.js
index 64db905..3a2fa67 100644
--- a/lib/llm/bedrock-response.js
+++ b/lib/llm/bedrock-response.js
@@ -38,11 +38,13 @@ class BedrockResponse {
     const cmd = this.#command
     if (cmd.isAi21() === true) {
       this.#completions = this.#parsedBody.completions.map((c) => c.data.text)
+      this.id = this.#parsedBody.id
     } else if (cmd.isClaude() === true) {
       // TODO: can we make this thing give more than one completion?
       this.#completions.push(this.#parsedBody.completion)
     } else if (cmd.isCohere() === true) {
       this.#completions = this.#parsedBody.generations.map((g) => g.text)
+      this.id = this.#parsedBody.id
     } else if (cmd.isTitan() === true) {
       this.#completions = this.#parsedBody.results.map((r) => r.outputText)
     }
diff --git a/lib/llm/chat-completion-message.js b/lib/llm/chat-completion-message.js
index 8f55620..916fc18 100644
--- a/lib/llm/chat-completion-message.js
+++ b/lib/llm/chat-completion-message.js
@@ -61,7 +61,7 @@ class LlmChatCompletionMessage extends LlmEvent {
     if (cmd.isTitan() === true || cmd.isClaude() === true) {
       this.id = `${this.id}-${index}`
     } else if (cmd.isAi21() === true || cmd.isCohere() === true) {
-      this.id = `${this.response.id}-${index}`
+      this.id = `${this.bedrockResponse.id}-${index}`
     }
   }
 }
```

File: lib/llm/chat-completion-summary.js [hunk 9]
```diff
@@ -39,17 +39,17 @@ class LlmChatCompletionSummary extends LlmEvent {
     this[uct] = parseInt(this.bedrockResponse.headers['x-amzn-bedrock-output-token-count'] || 0, 10)
     this[upt] = parseInt(this.bedrockResponse.headers['x-amzn-bedrock-input-token-count'] || 0, 10)
```
Comment (bizob2828) [inline lib/llm/chat-completion-summary.js:40]: this isn't using the new `this.bedrockResponse.inputTokencount`

File: tests/unit/llm/bedrock-response.tap.js [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+      return false
+    }
+  }
+
+  t.context.updatePayload = (payload) => {
+    t.context.response.output.body = new TextEncoder().encode(JSON.stringify(payload))
+  }
+})
+
+tap.test('non-conforming response is handled gracefully', async (t) => {
```
Comment (bizob2828) [inline tests/unit/llm/bedrock-response.tap.js:86]: when i run the tests locally and look at the coverage this for some reason isn't hitting the else statement where it defaults to 0

**Review by bizob2828 — CHANGES_REQUESTED** [review]
overall looks good just a few questions/issues. I noticed the addition of those unit tests it no longer runs the `test/unit/utils.tap.js` file when i run` npm run unit`

File: tests/unit/llm/bedrock-response.tap.js (hunk 10 above)
Comment (bizob2828) [reply tests/unit/llm/bedrock-response.tap.js:86]: oh i see why, it's cuz the test is removing all the headers. if we just do this, the branch is covered better

```js
tap.test('non-conforming response is handled gracefully', async (t) => {
  delete t.context.response.response.headers['x-amzn-bedrock-input-token-count']
  const res = new BedrockResponse(t.context)
  t.same(res.completions, [])
  t.equal(res.finishReason, undefined)
  //t.same(res.headers, undefined)
  t.equal(res.id, undefined)
  t.equal(res.inputTokenCount, 0)
  t.equal(res.requestId, 'aws-request-1')
  t.equal(res.statusCode, 200)
})
```

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**